packages/headless/src/utils/builder.ts

```diff
@@ -0,0 +1,43 @@
+import { computed, type ComputedRef, type HtmlHTMLAttributes, type Ref } from "vue";
+
+export type IteratedHeadlessElementFunc<T extends Record<string, unknown>> = (
+  opts: T,
+) => HtmlHTMLAttributes;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type HeadlessElementAttributes = HtmlHTMLAttributes | IteratedHeadlessElementFunc<any>;
+
+export type HeadlessElements = Record<
+  string,
+  HeadlessElementAttributes | ComputedRef<HeadlessElementAttributes>
+>;
+
+export type HeadlessState = Record<string, Ref>;
+
+export type HeadlessComposable<Elements extends HeadlessElements, State extends HeadlessState> = {
+  elements: Elements;
+  state: State;
+};
+
+/**
+ * We use this identity function to ensure the correct typings of the headless composables
+ */
+export const createBuilder = <P, Elements extends HeadlessElements, State extends HeadlessState>(
```

Comment:
If this is only used for typing, can't we just define the type when creating the composable like so?:
```ts
export const useComboBox: HeadlessComposable = () => {
  // ...
}
```

Reply:
Unfortunately type inference does not work with assigned types.
With an assigned type we would need to to declare all types upfront:
```
type Props = {...};
type Elements = {...};
type State = {...};
export const useComboBox: HeadlessComposable<Props, Elements, State> = () => {
  // ...
}
```

Reply:
Optionally we could use satisfies like
```ts
export const useComboBox = () => {
  // ...
} satisfies HeadlessComposable;
```

Like that all props will be inferred correctly and we ensure the type.
But then we can probably also keep the createBuilder